### DISCUSSION TITLE
Question: How do I generate client code

### DISCUSSION THREAD

**Discussion by chris-kruining** [thesayyn/protoc-gen-ts#109] [Ideas]
Hi,

I've just managed to get your plugin up and running as a replacement for `--js-out`. however I still generate client code with the grpc-web plugin, but that plugin has hardcoded to suffix all files with `_pb`, so that also means for the imports.

can I:
a) replace grpc-web with something else that generates my client stub?
b) convince this plugin to add the `_pb` suffix to imports (would require a patch I presume since [`replaceExtension(dependency, "");` ](https://github.com/thesayyn/protoc-gen-ts/blob/b47d74afc5b35cf13bd66a57b2f77c56debee212/src/index.js#L86) has a hardcoded empty string)

**Comment by chris-kruining** [comment]
nvm, I am straight up an idiot. I did not realize the plugin creates the client in the same file. old habits I suppose. Sorry to bother for nothing...

**Comment by thesayyn** [comment]
Hey, 

Recently I started working on #102 which will workaround grpc-web generation. currently, the clients are generated for node.

**Comment by chris-kruining** [comment]
> Hey,
> 
> Recently I started working on #102 which will workaround grpc-web generation. currently, the clients are generated for node.

awesome!! need any help? recently client-to-server streaming became a reality with fetch upload streaming. Do you reckon you could implement that as well? Or could I assist by doing that myself? I also have a "feature request" of sorts that implements async generators for streaming requests in either direction :D

Again, not afraid to get my hand dirty, so just tell me if I can assist :D

**Comment by thesayyn** [comment]
I like your enthusiasm. In that PR the generation part is complete but am trying to get an e2e test running. I tend to break things without noticing.

I believe that in order to support  CLIENT_STREAMING (possibly BIDI_STREAMING), you need to do some upstream work in grpc-web, tonic-web, and envoy-proxy which would be a lot of work I imagine.  I would love to see this landed though.

I don't know if they would ever want to use a platform API that is not widely supported by major browser vendors. 

Sometimes it is the resistance that holds you back instead of the work.

**Comment by chris-kruining** [comment]
hmm, but with a working implementation of client and bidi streaming, is grpc-web no long a thing of the past, wouldn't those 2 features make browsers compatible to the point it is just another grpc client? AFAIK that is why grpc-web exists, because the browser did not support upload streaming and is therefor incompatible with grpc.

So I would think that there is not much upstream issue as an envoy proxy for example could just speak grpc both ways instead of having to translate to grpc-web.

**Comment by thesayyn** [comment]
> AFAIK that is why grpc-web exists, because the browser did not support upload streaming and is therefor incompatible with grpc.

it exists because some browsers do not support HTTP/2 specific features such as frames.


> So I would think that there is not much upstream issue as an envoy proxy for example could just speak grpc both ways instead of having to translate to grpc-web.

envoy proxy could speak to the origin server any way it wants but it needs to understand what grpc-web says.

**Reply by chris-kruining** [reply]
> it exists because some browsers do not support HTTP/2 specific features such as frames.

wait, so streaming data != data frames? if I interpret https://developers.google.com/web/fundamentals/performance/http2#streams_messages_and_frames correctly this is the case right?

> envoy proxy could speak to the origin server any way it wants but it needs to understand what grpc-web says.

I know, my point is that grpc-web is not needed if the browser is capable of doing grpc. Which I suspect is now possible due to fetch upload streaming, and my interpretation of the above being correct.

**Reply by thesayyn** [reply]
> wait, so streaming data != data frames? if I interpret https://developers.google.com/web/fundamentals/performance/http2#streams_messages_and_frames correctly this is the case right?
> 
> >

yes, that is what I meant.

**Reply by thesayyn** [reply]
> I know, my point is that grpc-web is not needed if the browser is capable of doing grpc. Which I suspect is now possible due to fetch upload streaming, and my interpretation of the above being correct.

Now, this is getting interesting and yes that would be the case if the browser can speak native grpc protocol over h2. 
Afaik there were some other limitations like concurrent xhr/fetch requests. I believe that still exists with/without grpc-web.

**Comment by thesayyn** [comment]
See: https://github.com/envoyproxy/envoy/blob/9dd8ae3c69a9917d97a886ed03e1c010dcd9b098/source/extensions/filters/http/grpc_web/grpc_web_filter.cc
